kernel/interrupt.c

```diff
@@ -36,19 +36,22 @@ static const char * exception_names[] = {
 
 static void unknown_exception( int i, int code )
 {
-	unsigned vaddr, paddr;
+	unsigned vaddr, paddr, esp;
 
 	if(i==14) {
-		asm("mov %%cr2, %0" : "=r" (vaddr) );
-
-		// XXX Now that the process tracks the current data/stack size, we can more easily distinguish the reason for the fault.
-		int data = vaddr < (int)current->vm_data_size;
-		int stack = current->vm_data_size && pagetable_getmap(current->pagetable,vaddr+PAGE_SIZE,&paddr,0);
-
-		//Check for a valid mapping (which will result from violating the permissions on page), or that
-		//we are either accessing neither the stack nor the heap, or somehow both, and error if so
-
-		if (pagetable_getmap(current->pagetable,vaddr,&paddr,0) || !(data ^ stack)) {
+		asm("mov %%cr2, %0" : "=r" (vaddr) ); // virtual address trying to be accessed
+		asm("mov %%esp, %0" : "=r" (esp) ); // stack pointer
```

Comment:
You are on the right track, but this doesn't do quite what you think it does.
This extracts the *current* value of the stack pointer into the variable esp.
Try printing it out, and you will see that the current esp is somewhere in the
kstack for the process, which is used while running in kernel mode.  Your
code happens to work because this is below *all* possible user addresses.

What you want to extract is the value of the stack pointer as it appeared
just *before* the page fault.  That value (along with others) was pushed
onto the kstack when the interrupt occurred.

If you look at the structure `x86_stack`, it describes the interrupt stack layout (in reverse order).
The first thing pushed onto the stack is `ss` and the second is `esp`.  So, if you grab the second
item from the top of the kstack, it will be the former esp.  Print it out to verify: the value should
point way to the top of user memory: `0xffffffxx` or thereabouts.

A more elegant way is to declare a pointer `p` of type `x86_stack` and make it point to the top
of the kernel stack minus the size of `x86_stack` and then you can just reach in and grab `p->esp`

(Notice how the kernel stack is set up in`process_create` and `process_stack_reset`)

Reply:
Oh okay that makes sense. I will fix that.

Reply:
Looking at the code, it looks like:
`current->kstack_ptr = current->kstack_top - sizeof(struct x86_stack)` 
Then, we can say:
`esp = ((struct x86_stack *)current->kstack_ptr)->esp`
That will give the stack pointer before the interrupt, correct?

Reply:
That looks right to me, I think

Reply:
So I am trying to add some test cases that produce failures. The first scenario I am trying to get a failure for is accessing the stack where vaddr < esp. Here is my code for all the test cases (trying to fail is the last portion, Test 4):

``` c
int main() {
	// Test 1: Check if we can allocate a large 2D array on the stack
	int t_size = 500, i, j;
	int test_1_array_1 [t_size][t_size];
	int test_1_array_2 [t_size][t_size];
	for (i = 0; i < t_size; ++i) {
		for (j = 0; j < t_size; ++j) {
			test_1_array_1[i][j] = i%10; 
			test_1_array_2[i][j] = i%15; 
		}
	}
	printf_putstring("Test 1 passed!\n");
	flush();

	// Test 2: Check that allocating on stack does not interfere
	int test_2_array_1[10000];
	int * test_2_array_2 = malloc(10000);

	for (i = 0; i < 10000; ++i)
	{
		test_2_array_1[i] = i%10;
		test_2_array_2[i] = i%15;
	}

	free(test_2_array_2);

	printf_putstring("Test 2 passed!\n");
	flush();	

	// Test 3: Make sure calls to other functions works
	func1();
	int test_3_array[100000];
	for (i = 0; i < 100000; ++i) {
		test_3_array[i] = i%10;
	}
	
	printf_putstring("Test 3 passed!\n");
	flush();	

	// Test 4: Test for failure of malloc
	unsigned sp;
	asm( "mov %%esp, %0" : "=rm" ( sp ));
	int failure = 1;
	int * fail_array_1;
	int * temp;
	int fail_array_2[10000];

	if (failure == 1) {
		temp = (int *)((void *)fail_array_2 - 2000131);
		temp[0] = 1;
		printf("Pointer address %u\n", sp);
		printf("Pointer address %u\n", fail_array_2);
		printf("Pointer address %u\n", &temp[0]);
	}

	printf_putstring("Fail passed!\n");
	flush();
	return 0;
}

```

"- 2000131" is from a calculation I did based on the value of "fail_array_1 - sp". However, this does not raise an exception. So, my question is, why unknown_exception in interrupt.c does not even get called when `temp[0] = 1 ` is executed and temp is less than sp?

Reply:
> Looking at the code, it looks like:
> `current->kstack_ptr = current->kstack_top - sizeof(struct x86_stack)`
> Then, we can say:
> `esp = ((struct x86_stack *)current->kstack_ptr)->esp`
> That will give the stack pointer before the interrupt, correct?

Close, except that current->kstack_ptr changes over time as the kstack gets used.
Just subtract from kstack_top and use that value.